### PR TITLE
Dref Superticket

### PR DESCRIPTION
**Pull request by k9845**
Addresses
 - https://github.com/IFRCGo/go-web-app/issues/352
## Checklist
Things that should succeed before merging.

- [ ] Updated/ran unit tests
- [ ] Updated CHANGELOG.md

## Release

If there is a version update, make sure to tag the repository with the latest version.

### CODE REVIEW

**Comment by szabozoltan69** [comment]
Before the merge to develop I would like to (ask you to) clean up the migration files.
As I see there were changes regarding (almost) the same fields, so it would be fine to use only the latest changes.
These migration files always run at test time, so they increases the deployment time always.
(Also we planned to unifiy migration files, didn't we, @thenav56 ?)

**Comment by thenav56** [comment]
> Before the merge to develop I would like to (ask you to) clean up the migration files. As I see there were changes regarding (almost) the same fields, so it would be fine to use only the latest changes. These migration files always run at test time, so they increases the deployment time always. (Also we planned to unifiy migration files, didn't we, @thenav56 ?)

Totally agree with you there.
We can also explore with a CI check, where we show a warning if there are lots of Django migrations.
For existing migrations, we can look into Django squash migration https://docs.djangoproject.com/en/4.2/topics/migrations/#squashing-migrations

**Comment by k9845** [comment]
@szabozoltan69 @thenav56 created a squashed migration file.

**Review by szabozoltan69 — CHANGES_REQUESTED** [review]
Please change DISPALCEMENT to DISPLACEMENT.

**Comment by szabozoltan69** [comment]
Still checking...
(Found a word Diaster (instead of Disaster) in dref/models.py, but still not finished with the check. Need some time still...)

File: dref/models.py [hunk 1]
```diff
@@ -18,27 +18,27 @@
...
-        NATIONAL_SOCIETY_EOC = "national_society_eoc", _("National Society EOC")
         SHELTER_HOUSING_AND_SETTLEMENTS = "shelter_housing_and_settlements", _("Shelter, Housing And Settlements")
         LIVELIHOODS_AND_BASIC_NEEDS = "livelihoods_and_basic_needs", _("Livelihoods And Basic Needs")
+        MULTI_PURPOSE_CASH = "multi_purpose_cash", _("Multi Purpose Cash")
         HEALTH = "health", _("Health")
         WATER_SANITATION_AND_HYGIENE = "water_sanitation_and_hygiene", _("Water, Sanitation And Hygiene")
         PROTECTION_GENDER_AND_INCLUSION = "protection_gender_and_inclusion", _("Protection, Gender And Inclusion")
         EDUCATION = "education", _("Education")
-        MIGRATION = "migration", _("Migration")
+        MIGRATION_AND_DISPLACEMENT = "migration_and_displacement", _("Migration And Displacement")
```
Comment (thenav56) [inline dref/models.py:28]: We will need to migrate data as well.

File: dref/models.py [hunk 2]
```diff
@@ -78,20 +78,19 @@ class IdentifiedNeed(models.Model):
...
         SHELTER_HOUSING_AND_SETTLEMENTS = "shelter_housing_and_settlements", _("Shelter Housing And Settlements")
         LIVELIHOODS_AND_BASIC_NEEDS = "livelihoods_and_basic_needs", _("Livelihoods And Basic Needs")
+        MULTI_PURPOSE_CASH_GRANTS = "multi_purpose_cash_grants", _("Multi purpose cash grants")
         HEALTH = "health", _("Health")
         WATER_SANITATION_AND_HYGIENE = "water_sanitation_and_hygiene", _("Water, Sanitation And Hygiene")
         PROTECTION_GENDER_AND_INCLUSION = "protection_gender_and_inclusion", _("Protection, Gender And Inclusion")
         EDUCATION = "education", _("Education")
-        MIGRATION = "migration", _("Migration")
-        MULTI_PURPOSE_CASH_GRANTS = "multi_purpose_cash_grants", _("Multi purpose cash grants")
+        MIGRATION_AND_DISPLACEMENT = "migration_and_displacement", _("Migration And Displacement")
```
Comment (thenav56) [inline dref/models.py:86]: We will need to migrate data as well.

File: dref/models.py [hunk 3]
```diff
@@ -18,27 +18,27 @@
...
         WATER_SANITATION_AND_HYGIENE = "water_sanitation_and_hygiene", _("Water, Sanitation And Hygiene")
         PROTECTION_GENDER_AND_INCLUSION = "protection_gender_and_inclusion", _("Protection, Gender And Inclusion")
         EDUCATION = "education", _("Education")
-        MIGRATION = "migration", _("Migration")
+        MIGRATION_AND_DISPLACEMENT = "migration_and_displacement", _("Migration And Displacement")
         RISK_REDUCTION_CLIMATE_ADAPTATION_AND_RECOVERY = "risk_reduction_climate_adaptation_and_recovery", _(
             "Risk Reduction, Climate Adaptation And Recovery"
         )
-        COMMUNITY_ENGAGEMENT_AND_ACCOUNTABILITY = "community_engagement_and _accountability", _(
+        COMMUNITY_ENGAGEMENT_AND_ACCOUNTABILITY = "community_engagement_and_accountability", _(
```
Comment (thenav56) [inline dref/models.py:32]: We will need to migrate data as well.

File: dref/models.py [hunk 4]
```diff
@@ -18,27 +18,27 @@
...
     class Title(models.TextChoices):
-        NATIONAL_SOCIETY_READINESS = "national_society_readiness", _("National Society Readiness")
-        ASSESSMENT = "assessment", _("Assessment")
-        COORDINATION = "coordination", _("Coordination")
-        RESOURCE_MOBILIZATION = "resource_mobilization", _("Resource Mobilization")
-        ACTIVATION_OF_CONTINGENCY_PLANS = "activation_of_contingency_plans", _("Activation Of Contingency Plans")
-        NATIONAL_SOCIETY_EOC = "national_society_eoc", _("National Society EOC")
         SHELTER_HOUSING_AND_SETTLEMENTS = "shelter_housing_and_settlements", _("Shelter, Housing And Settlements")
         LIVELIHOODS_AND_BASIC_NEEDS = "livelihoods_and_basic_needs", _("Livelihoods And Basic Needs")
+        MULTI_PURPOSE_CASH = "multi_purpose_cash", _("Multi Purpose Cash")
```
Comment (thenav56) [inline dref/models.py:23]: We will need to migrate data as well.

File: dref/models.py [hunk 5]
```diff
@@ -78,20 +78,19 @@ class IdentifiedNeed(models.Model):
...
         PROTECTION_GENDER_AND_INCLUSION = "protection_gender_and_inclusion", _("Protection, Gender And Inclusion")
         EDUCATION = "education", _("Education")
-        MIGRATION = "migration", _("Migration")
-        MULTI_PURPOSE_CASH_GRANTS = "multi_purpose_cash_grants", _("Multi purpose cash grants")
+        MIGRATION_AND_DISPLACEMENT = "migration_and_displacement", _("Migration And Displacement")
         RISK_REDUCTION_CLIMATE_ADAPTATION_AND_RECOVERY = "risk_reduction_climate_adaptation_and_recovery", _(
             "Risk Reduction, Climate Adaptation And Recovery"
         )
-        COMMUNITY_ENGAGEMENT_AND_ACCOUNTABILITY = "community_engagement_and _accountability", _(
+        COMMUNITY_ENGAGEMENT_AND_ACCOUNTABILITY = "community_engagement_and_accountability", _(
```
Comment (thenav56) [inline dref/models.py:90]: We will need to migrate data as well.

File: dref/models.py [hunk 6]
```diff
@@ -78,20 +78,19 @@ class IdentifiedNeed(models.Model):
...
             "Community Engagement And Accountability"
         )
-        ENVIRONMENT_SUSTAINABILITY = "environment_sustainability ", _("Environment Sustainability")
-        SHELTER_CLUSTER_COORDINATION = "shelter_cluster_coordination", _("Shelter Cluster Coordination")
```
Comment (thenav56) [inline dref/models.py:94]: We will need to migrate data as well.

File: dref/serializers.py [hunk 7]
```diff
@@ -346,15 +357,16 @@ class DrefSerializer(NestedUpdateMixin, NestedCreateMixin, ModelSerializer):
     country_details = MiniCountrySerializer(source="country", read_only=True)
     district_details = MiniDistrictSerializer(source="district", read_only=True, many=True)
     assessment_report_details = DrefFileSerializer(source="assessment_report", read_only=True)
-    supporting_document_details = DrefFileSerializer(read_only=True, source="supporting_document")
+    supporting_document_details = DrefFileSerializer(source="supporting_document", read_only=True)
```
Comment (thenav56) [inline dref/serializers.py:360]: Looks like we need to add additional parameters `required=False, allow_null=True`  to specify nullable fields.

File: dref/test_views.py [hunk 8]
```diff
@@ -311,6 +311,7 @@ def test_event_date_in_dref(self):
         url = "/api/v2/dref/"
         self.client.force_authenticate(self.user)
         response = self.client.post(url, data, format="json")
+        print(response.content)
```
Comment (thenav56) [inline dref/test_views.py:314]:
```suggestion
```

**Review by thenav56 — CHANGES_REQUESTED** [review]
Looks good. Other than typos, we need to make sure to migrate the legacy data.

For the migration, let's discuss further. Maybe we can merge everything into a single file before merging this to develop.

File: dref/models.py (hunk 6 above)
Comment (thenav56) [reply dref/models.py:94]: @frozenhelium @udaynwa 
Can you confirm the behaviour here?

Currently, this enums are used by title which are non-nullable field. We have 3 options.

- Ignore... Will raise error on runtime
- Migrate to another value
- Delete IdentifiedNeed objects related to this value

Let's figure out the best case.

**Review by thenav56 — APPROVED** [review]

File: dref/models.py (hunk 6 above)
Comment (frozenhelium) [reply dref/models.py:94]: Let's do the option (c) "Delete `IdentifiedNeed` objects related to that value"

Comment (k9845) [reply dref/models.py:94]: Removed `IdentifiedNeed`  related objects.

**Review by thenav56 — APPROVED** [review]

**Comment by thenav56** [comment]
@szabozoltan69 All good here. Can you review again and merge this? Thanks

**Comment by szabozoltan69** [comment]
What a nice and gentle solution!
reverse_code=migrations.RunPython.noop

**Review by szabozoltan69 — APPROVED** [review]

**Pull request merged**